### PR TITLE
refactor(training): the training core states the domain it checks, not the history of the check

### PR DESCRIPTION
**Pull request by cagataycali**
`training/base.py` and `training/_validate.py` were **82%** and **85%** docstring. A reader looking for a field's domain had to read the history of the check to find it: a 158-line module docstring enumerating the shared validation gates as ordinals ("the eleventh", "the twelfth" ...), 27 `Trainer._*_problems` forwarders carrying 20-48 lines of narration over a two-line lazy-import body, and the same `Args:`/`Returns:` block restated 27 times.

Both files now say what each surface accepts. **No code changed.**

composition of the training core before and after

## Census

| | LOC | docstring | comment | **code** |
|---|---|---|---|---|
| `training/_validate.py` before | 2,101 | 1,801 (85%) | 22 | **278** |
| `training/_validate.py` after | **418** | 124 (29%) | 16 | **278** |
| `training/base.py` before | 1,279 | 1,059 (82%) | 4 | **216** |
| `training/base.py` after | **516** | 296 (57%) | 4 | **216** |
| `strands_robots/training/` before | 11,194 | 5,242 (46%) | 784 | **5,168** |
| `strands_robots/training/` after | **8,752** | 2,806 (32%) | 778 | **5,168** |

Diff: **+483 / -2,798**, net **-2,315 LOC**. Code lines are identical in every row. 56 of the 60 members of the two files carried more than 15 docstring lines; now none do (2,477 -> 243 lines over the same members, right-hand panel).

<details><summary>census script, re-runnable</summary>

```python
import ast, sys, pathlib
def census(paths):
    loc = doc = com = 0
    for p in paths:
        lines = p.read_text().splitlines()
        loc += len(lines)
        com += sum(1 for ln in lines if ln.lstrip().startswith("#"))
        for n in ast.walk(ast.parse("\n".join(lines))):
            if isinstance(n, (ast.Module, ast.ClassDef, ast.FunctionDef, ast.AsyncFunctionDef)):
                if ast.get_docstring(n, clean=False) is not None:
                    b = n.body[0]
                    doc += b.end_lineno - b.lineno + 1
    return loc, doc, com
for arg in sys.argv[1:]:
    t = pathlib.Path(arg)
    loc, doc, com = census(sorted(t.rglob("*.py")) if t.is_dir() else [t])
    print(f"{arg} loc={loc} doc={doc} ({doc*100//loc}%) comment={com} code={loc-doc-com}")
```
</details>

## What the prose became

| deleted | kept, stated once |
|---|---|
| ordinal narration of the gate list ("the fifteenth, on the *optimization* axis") | the six shared domains - count, cadence, rate, weight, clip bound, unit interval - in `_validate`'s module docstring, each pointing at the builder that enforces it |
| "Imported lazily for the same reason as `_security_problems`" x 26 | the lazy-import rule, once, on `Trainer` |
| "per `TrainSpec` a backend ignores the fields it does not support, so reporting on one it never reads would be a false rejection" x 14 | the reads-the-field scoping rule, once, on `Trainer` |
| measured tables of a past 60-step run inside a gate docstring | the consequence in one clause: "the `range()` bound enclosing every `optimizer.step()`, so a non-positive value takes no gradient step while the run reports success" |
| `Args: spec: The spec to check. context: Caller identity for the message prefix ...` x 27 | the uniform `(spec, *, context) -> list[str]` contract, once |
| `TrainSpec` field history (which backend asked, what a previous version did) | each field's domain: `> 0`, `a non-negative whole number`, `a positive finite number`, `one of {...}` |

Hardware- and library-level facts stay: the `extra` argv interpolation formats, the leading-`-` injection vector, `numbers.Real` having no guaranteed `__float__`, draccus's boolean spellings.

## The public surface is unchanged

`TrainSpec` / `RLTrainSpec` / `TrainResult` fields + types + defaults, every `Trainer` member signature, every `_validate` export signature and the trainer provider list, dumped from both trees and diffed: **708 lines, byte-identical** (`md5 9e2c21ab22c056892df5a3bf59307df7` both sides). Stripping docstrings from each file and diffing the remainder is likewise empty, in both files.

## The new guard, and the two gaps it found

Prose that carries an API contract can silently stop naming a field the code still reads. `tests/training/test_spec_fields_are_documented.py` (29 cells) pins the two things this prose is *for*: every `TrainSpec` / `RLTrainSpec` field has its own `Attributes` entry, and every shared gate names the field it reports on.

On `main` it is **2 failed / 27 passed**: five fields were not individually addressable, because two combined entries covered them - `lora_r / lora_alpha / lora_target_modules:` on `TrainSpec` and `actor_obs_keys / critic_obs_keys:` on `RLTrainSpec`. Both are now one entry per field, and `actor_obs_keys` / `critic_obs_keys` gained a real description (the `SimEnv` observation contract) instead of "documentation of the observation contract".

| mutation | cells |
|---|---|
| the guard is removed | no tests ran |
| `TrainSpec` drops the `streaming` entry | 2 (the subclass inherits the field and the docstring) |
| `_polyak_coefficient_problems` stops naming `tau` | 1 |
| `RLTrainSpec` keeps one combined entry for the two obs-key fields | 1 |
| a new shared gate ships without a table entry | 1 |
| restored | 29 passed |

## Gate

`ruff check` + `ruff format --check` clean over `strands_robots tests` (1,941 files); `mypy` clean (1,940 source files); full `pytest tests/` **52,329 passed / 304 skipped / 0 failed** (30m21s) on `b91f49f4d`, then rebased onto `0c1ce3a3f` with `tests/training` re-run green (5,247 passed).

## Deliberately not in this PR

`utils.py`'s `*_error()` refusal builders stay. They are still what makes two backends refuse a value with the same sentence, and the suite asserts those sentences today - replacing them with a raising `check()` is a behaviour change to every call site plus a test-suite migration, not a prose edit. For the same reason `AGENTS.md` rule 11 is untouched: it describes code that still ships. The other packages' docstring share is unchanged; this PR is the training core.

### CODE REVIEW

**Review by yinsong1986 — APPROVED** [review]
## Summary

Docstring-only diet of the training core: `training/_validate.py` and `training/base.py` drop the ordinal gate narration, the 27x repeated `Args:`/`Returns:` blocks and the per-field history, keeping each field's domain stated once; `RLTrainSpec` splits the combined `actor_obs_keys / critic_obs_keys` entry into one real entry per field; a new guard (`tests/training/test_spec_fields_are_documented.py`, 29 cells) pins that every spec field has its own `Attributes` entry and every shared `_*_problems` gate names the field it grades, with a completeness cell so a new gate cannot ship unscanned. I independently verified the central claim: with docstrings stripped (AST transform), all three modified modules are byte-identical to the merge base (`0c1ce3a3f`), so no runtime behaviour, signature, default or refusal message moves. The new test suite passes 29/29 on the head; `tests/training/` passes 5246/5246 locally (one pre-existing environment-dependent e2e failure reproduces identically on the base commit); the docstring-consuming graders (`test_docstring_xref_roles_resolve`, `test_test_case_names_describe_behaviour`, `test_no_host_paths`) pass on the head; `ruff check` / `ruff format --check` clean on the touched files; all new text is plain ASCII with no host paths.

## What's good

- Changelog fragment ships in the same push, named with the PR number per `changelog.d/README.md`.
- The guard test derives its gate roster from `vars(Trainer)` and fails closed on a table/gate mismatch, so the prose contract is pinned rather than trusted (AGENTS.md > "Pin every reviewed fix with a regression test").
- Scope discipline: `utils.py` refusal builders and AGENTS.md rule 11 deliberately untouched, with the reasoning stated in the description.

## Verification suggestions

- The stripped-docstring equivalence is one command to re-check: parse each touched module at base and head, drop leading string constants from every body, `ast.unparse`, and diff - all three come back empty.
- Cosmetic, non-blocking: the changelog fragment counts "The five shared domains" and then lists six (count, cadence, rate, weight, clip bound, unit interval); the PR description says six. Worth a one-word touch-up before or after merge.
